### PR TITLE
Dropout optimized for p == 0

### PR DESCRIPTION
**Pull request by ivpopov**
Dropout optimized for p == 0 not to perform any operations

### CODE REVIEW

File: Dropout.lua [hunk 1]
```diff
@@ -14,34 +14,42 @@ function Dropout:__init(p,v1,inplace)
...
+          self.noise:bernoulli(1-self.p)
+          if self.v2 then
+             self.noise:div(1-self.p)
+          end
+          self.output:cmul(self.noise)
+      elseif not self.v2 then
+          self.output:mul(1-self.p)
+      end
+   else
+      self.output = input
```
Comment (soumith) [inline Dropout.lua]: this does not respect in-place / out-of-place.
Instead, remove the if condition in line 17, remove this else condition and add:
if self.p >= 0 then
... line 18 to 22
end

File: Dropout.lua [hunk 2]
```diff
@@ -14,34 +14,42 @@ function Dropout:__init(p,v1,inplace)
...
+         end
+         self.gradInput:cmul(self.noise) -- simply mask the gradients with the noise vector
       else
-         self.gradInput:resizeAs(gradOutput):copy(gradOutput)
+         error('backprop only defined while training')
       end
-      self.gradInput:cmul(self.noise) -- simply mask the gradients with the noise vector
    else
-      error('backprop only defined while training')
+      self.gradInput = gradOutput
```
Comment (soumith) [inline Dropout.lua]: same comments as in updateOutput. Doesn't respect in-place / out-of-place. instead reuse the logic in line 42-46

**Comment by ivpopov** [comment]
Changed

**Comment by soumith** [comment]
Thanks Ivan.

**Pull request merged**